### PR TITLE
Update json schema to be generated from typescript + fixes 

### PR DESCRIPTION
**Pull request by Mearman**
- Switch to generating the JSON schema directly from the Typescript types.
- Adds testing of the generated schema against all boards in the repository

### CODE REVIEW

**Comment by PaulKinlan** [comment]
Hmm - I'd need to understand the implications a bit more. As I read it `identifier` is used to name nodes in `from` and `to` which would mean that any node name could start with $ in an edge, not just error.

File: .changeset/slow-cougars-promise.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@google-labs/breadboard-schema": minor
```
Comment (aomarks) [inline .changeset/slow-cougars-promise.md:2]: patch (bugfix)

**Review by aomarks — CHANGES_REQUESTED** [review]

File: .changeset/slow-cougars-promise.md (hunk 1 above)
Comment (Mearman) [reply .changeset/slow-cougars-promise.md:2]:
```suggestion
"@google-labs/breadboard-schema": patch
```

File: packages/schema/breadboard.schema.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
   "$defs": {
     "identifier": {
       "type": "string",
-      "pattern": "^[a-zA-Z_][a-zA-Z0-9_-]*$"
+      "pattern": "^[\\$a-zA-Z_][a-zA-Z0-9_-]*$"
```
Comment (Mearman) [inline packages/schema/breadboard.schema.json:9]:
```suggestion
      "pattern": "^(\\$error)|[a-zA-Z_][a-zA-Z0-9_-]*$"
```

**Comment by Mearman** [comment]
> Hmm - I'd need to understand the implications a bit more. As I read it `identifier` is used to name nodes in `from` and `to` which would mean that any node name could start with $ in an edge, not just `$error`.

good point @PaulKinlan
updated: https://github.com/breadboard-ai/breadboard/pull/571/files#r1475193072

File: packages/schema/breadboard.schema.json (hunk 2 above)
Comment (Mearman) [reply packages/schema/breadboard.schema.json:9]: I know the parentheses at the start are superfluous, it just makes it a bit more clear

**Comment by aomarks** [comment]
Forgive my ignorance, is $error a special kind of port? Do all nodes implicitly have an $error port?

**Comment by dglazkov** [comment]
> Forgive my ignorance, is $error a special kind of port? Do all nodes implicitly have an $error port?

yes! Doh! I need to add that to inspector API, too. The `$error` port is a way to catch errors from nodes.

**Comment by Mearman** [comment]
hah. I've seen it enough in error outputs

File: packages/schema/breadboard.schema.json [hunk 3]
```diff
@@ -6,7 +6,7 @@
   "$defs": {
     "identifier": {
       "type": "string",
-      "pattern": "^[a-zA-Z_][a-zA-Z0-9_-]*$"
+      "pattern": "^(\\$error)|[a-zA-Z_][a-zA-Z0-9_-]*$"
```
Comment (aomarks) [inline packages/schema/breadboard.schema.json:9]: Still not sure this is right, because the "identifier" def is used in a bunch of places here, like node IDs. It shouldn't be valid to have a node ID called `$error`. So I think we need to factor out separate $defs, one for node IDs, one for port IDs (and one for graph IDs?).

**Review by aomarks — CHANGES_REQUESTED** [review]

File: packages/schema/breadboard.schema.json (hunk 3 above)
Comment (Mearman) [reply packages/schema/breadboard.schema.json:9]: can do. and it would be neater than in the regex

File: .github/workflows/labs-prototypes-ci.yml [hunk 4]
```diff
@@ -11,13 +11,13 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        node-version: [20.x]
+        node-version: [21.x]
```
Comment (dglazkov) [inline .github/workflows/labs-prototypes-ci.yml:14]: Why was this necessary? Just curious. Our docs say Node >= v19, but the bots run on 21.

**Review by dglazkov — APPROVED** [review]
LGTM, but would love @aomarks to have a look as well. I am not as familiar with the `schema` package yet.

File: .github/workflows/labs-prototypes-ci.yml (hunk 4 above)
Comment (Mearman) [reply .github/workflows/labs-prototypes-ci.yml:14]: tests simply wouldn't pass `¯\_(ツ)_/¯`

File: packages/schema/__tests__/util/ascendToPackageDir.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import fs from "fs";
```
Comment (aomarks) [inline packages/schema/__tests__/util/ascendToPackageDir.ts:1]: Let's use snake_case for filenames

In general, let's follow https://google.github.io/styleguide/tsguide.html

File: packages/schema/__tests__/util/ascendToPackageDir.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import path from "path";
+
+export function ascendToPackageDir(packageName: string = "breadboard-ai") {
+  let directory = process.cwd();
```
Comment (aomarks) [inline packages/schema/__tests__/util/ascendToPackageDir.ts:5]: Avoid cwd, better to use import.meta.url to make paths relative to the current file. The user controls cwd so it can be anything, but import.meta.url will always be reliable.

File: packages/schema/__tests__/util/getBoardFiles.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import fs from "fs";
```
Comment (aomarks) [inline packages/schema/__tests__/util/getBoardFiles.ts:1]: Missing license

File: packages/schema/breadboard.schema.json [hunk 8]
```diff
@@ -1,178 +1,299 @@
...
-      }
-    },
-    "args": {
-      "description": "Arguments that are passed to the graph, useful to bind values to lambdas.",
-      "type": "object",
-      "additionalProperties": true
-    }
-  }
+	"$id": "https://raw.githubusercontent.com/breadboard-ai/breadboard/@google-labs/breadboard-schema@1.2.0/packages/schema/breadboard.schema.json",
+	"$schema": "http://json-schema.org/draft-07/schema#",
```
Comment (aomarks) [inline packages/schema/breadboard.schema.json:3]: I used the $schema version at https://json-schema.org/learn/getting-started-step-by-step#create-a-schema-definition, I don't actually know about the different versions, but could you confirm that this new one we've switched to is the best/latest one to use?

File: packages/schema/package.json [hunk 9]
```diff
@@ -22,29 +22,56 @@
...
-    "test": {
+    "test:js": {
+      "command": "node --test-reporter=spec --test ./dist/__tests__/**/*.test.js",
+      "dependencies": [
+        "build"
+      ],
+      "files": [
+        "breadboard.schema.json",
+        "package.json",
+        "dist/**/*",
```
Comment (aomarks) [inline packages/schema/package.json:55]: Don't think we want to include the dist/ input files here, nor the *.ts files. The `build` dependency is sufficient for that.

File: packages/schema/package.json [hunk 10]
```diff
@@ -56,32 +83,38 @@
...
-        "src/",
+        "package.json",
+        "src/**/*.ts",
         "tsconfig.json",
+        "__tests__/**/*.ts",
         "../../core/tsconfig/base.json"
       ],
       "output": [
-        "dist/",
+        "dist/**",
```
Comment (aomarks) [inline packages/schema/package.json:93]: dist/** and dist/ are the same (wireit uses the same semantics as npm `files` array, so if you give it a folder, it's implicitly recursive)

File: packages/schema/package.json [hunk 11]
```diff
@@ -56,32 +83,38 @@
...
         "tsconfig.tsbuildinfo"
       ]
     },
-    "build:update-json-schema": {
-      "command": "node dist/scripts/update-json-schema.js",
-      "dependencies": [
-        "build:tsc"
+    "build:schema": {
+      "command": "tsx src/scripts/generate-json-schema.ts",
```
Comment (aomarks) [inline packages/schema/package.json:98]: In general I'd like us to just do tsc -> node. Having a bunch of ways to do tsc compilation seems unnecessary. So, this should depend on buid:tsc, and should just invoke node on the ts files.

File: .changeset/green-ants-dream.md [hunk 12]
```diff
@@ -0,0 +1,5 @@
+---
+"@google-labs/breadboard-schema": minor
```
Comment (aomarks) [inline .changeset/green-ants-dream.md:2]: I think this might be major, because the JSON schema could have breaking changes in it. (I sort of regret jumping to 1.0.0 already now, since I think we'll have a lot of breaking changes, but since we'll move to the @breadboard-ai NPM org at some point, it doesn't really matter, we could reset to 0).

**Review by aomarks — CHANGES_REQUESTED** [review]
+1 to the idea of generating the JSON schema from the TypeScript so that they are in sync.

Could you change the PR title and description to match what this PR now does?

File: packages/schema/breadboard.schema.json (hunk 8 above)
Comment (Mearman) [reply packages/schema/breadboard.schema.json:3]: the majority with useful features are still draft. 2020 is newer. but still unsupported in many places. which was the result of a lot of my issues before. draft 07 is a lot more established

File: packages/schema/src/scripts/generate-json-schema.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env npx -y tsx
```
[the remark this answers is not in the document]
Comment (Mearman) [reply packages/schema/src/scripts/generate-json-schema.ts:1]:
```suggestion
```

File: packages/schema/__tests__/util/ascendToPackageDir.ts (hunk 5 above)
Comment (aomarks) [reply packages/schema/__tests__/util/ascendToPackageDir.ts]: Oh, yeah sorry, let's do kebab-case for filenames.

File: packages/schema/src/graph.ts [hunk 14]
```diff
@@ -173,6 +174,31 @@ export type GraphIdentifier = string;
  */
 export type SubGraphs = Record<GraphIdentifier, GraphDescriptor>;
 
+/**
+ * Properties used for testing the foundational traversal tests
+ */
+export type TestProperties = {
```
[the remark this answers is not in the document]
Comment (aomarks) [reply packages/schema/src/graph.ts:180]: Yeah unless we want users to use these fields, I don't think they should be here.

BTW if we convert from types to interfaces, then it will be possible to use https://www.typescriptlang.org/docs/handbook/declaration-merging.html#merging-interfaces to add fields just for our own tests.

(But I wouldn't do that conversion in this PR, I tried and it actually caused a bunch of build errors).

[the remark this answers is not in the document]
Comment (Mearman) [reply packages/schema/src/graph.ts:180]: I do love merged declarations and exports.

File: packages/schema/src/scripts/generate-schema-id.ts [hunk 15]
```diff
@@ -0,0 +1,16 @@
+import path from "path";
+import packageJson from "../../package.json";
+
+export function generateSchemaId() {
+  const PACKAGE_ROOT = process.cwd();
```
[the remark this answers is not in the document]
Comment (Mearman) [reply packages/schema/src/scripts/generate-schema-id.ts:11]: this one I neeed specifically relative to the package root. though I will use the `ascendToPackageDir` function I made

File: .github/workflows/labs-prototypes-ci.yml [hunk 16]
```diff
@@ -10,8 +10,11 @@ jobs:
   build:
     runs-on: ubuntu-latest
     strategy:
+      fail-fast: false
```
Comment (Mearman) [inline .github/workflows/labs-prototypes-ci.yml:13]:
```suggestion
```

**Review by dglazkov — APPROVED** [review]
LGTM.

**Comment by dglazkov** [comment]
I must not understand how PR machinery works. Why does it still show as "blocked"?

**Comment by Mearman** [comment]
@dglazkov when there have been requested changes, the requester of the changes has to give approval again

**Comment by aomarks** [comment]
> without the change, this is not valid

Could you update the PR description to match what the PR does?

File: packages/schema/src/graph.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (aomarks) [reply packages/schema/src/graph.ts]: I still see the testing stuff in the PR, but this comment was resolved? We shouldn't put our internal testing stuff in public interfaces.

[the remark this answers is not in the document]
Comment (dglazkov) [reply packages/schema/src/graph.ts]: I agree. I would rather make the schema less strict, allowing extra properties -- since that's likely going to happen.

File: packages/schema/src/scripts/generate-json-schema.ts [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+import fs from "fs";
+import path from "path";
+
+const filePath = path.resolve("src/graph.ts");
```
[the remark this answers is not in the document]
Comment (aomarks) [reply packages/schema/src/scripts/generate-json-schema.ts:12]: This was also resolved, but seems to not have been changed?

I haven't gone through all the comments, but please only resolve comments if you've addressed them or explained why you don't agree we should address them (actually generally best not to click resolve until there's agreement between author/reviewer).

[the remark this answers is not in the document]
Comment (Mearman) [reply packages/schema/src/scripts/generate-json-schema.ts:12]: I thought I had. I shall check if it was lost in one of the rebases

File: packages/schema/src/graph.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Mearman) [reply packages/schema/src/graph.ts]: as a stop-gap before looking at migrating to interfaces I updated it so it's no longer exported and uses the `@deprecated` flag to highlight that it should be used
https://github.com/breadboard-ai/breadboard/pull/571/files#diff-cd98c3c8ef3db059cf8ac5041c98a0e771600bd33fc580fd4c78e37c0d36336aR199-R204

**Comment by Mearman** [comment]
update that updated that path resolution @aomarks

https://github.com/breadboard-ai/breadboard/pull/571/files#diff-ab6fdb0fb6720b120ffdae17c601c0dcf833dea31471429f9499c1ba710f2d79R14-R20

**Comment by dglazkov** [comment]
Let's land this -- found that I needed it today.

File: packages/schema/src/graph.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Mearman) [reply packages/schema/src/graph.ts]: happy to mark this resolved?

[the remark this answers is not in the document]
Comment (aomarks) [reply packages/schema/src/graph.ts]: Ok. Here's some inspiration for why I think we should be careful about keeping private implementation stuff out of our public interfaces: https://www.hyrumslaw.com/

**Review by aomarks — APPROVED** [review]

File: packages/schema/src/graph.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Mearman) [reply packages/schema/src/graph.ts]: Hah, I enjoy the XKCD for illustration. I'm certainly going to reference Hyrum's Law as it's something I've thought about but not been able to articulate well in the past.
But I feel that having the schema for the testing values and having them clearly marked as such aligns with Hyrum's Law better than having them not checked/annotated or marked as invalid?
